### PR TITLE
settings page

### PR DESCRIPTION
**Pull request by penk85**
WPC-12: Settings page

### CODE REVIEW

File: packages/api/src/db/setting.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+export type Setting<T = unknown> = {
+  id: string
+
+  name: string
+  value: T | null
+  settingRestriction?: SettingRestriction
+}
+
+export interface SettingInput<T> {
```
Comment (antkiewiczk) [inline packages/api/src/db/setting.ts:9]: maybe we can reuse SettingInput in create and update methods? similar as in e.g. subscriptions or payments

File: packages/api/src/graphql/query.private.ts [hunk 2]
```diff
@@ -1141,6 +1142,29 @@ export const GraphQLQuery = new GraphQLObjectType<undefined, Context>({
...
+    // Setting
+    // ======
+
+    setting: {
+      type: GraphQLSetting,
+      args: {id: {type: GraphQLID}, name: {type: GraphQLString}},
+      resolve(root, {id, name}, {authenticate, loaders}) {
+        // TODO authenticate
+
+        if ((id == null && name == null) || (id != null && name != null)) {
```
Comment (antkiewiczk) [inline packages/api/src/graphql/query.private.ts:1155]: We might think about using some kind of validations tool like Joi (https://joi.dev/api/?v=17.6.0) - something to think about for the future

File: packages/api-db-mongodb/src/db/schema.ts [hunk 3]
```diff
@@ -435,3 +438,11 @@ export interface DBMailLog {
   mailProviderID: string
   mailData?: string
 }
+
+export interface DBSetting<T> {
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/api-db-mongodb/src/db/schema.ts:442]: I think so, unless we decide `value` can be type `any`. It would be string, boolean, or number depending on the case.

File: packages/api-db-mongodb/src/db/setting.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      },
+      {returnOriginal: false}
+    )
+    if (!value) return null
+
+    const {_id: outID, ...setting} = value
+    return {id: outID, ...setting}
+  }
+
+  async deleteSetting({id}: DeleteSettingArgs): Promise<string | null> {
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/api-db-mongodb/src/db/setting.ts:70]: That sounds like a good idea. I can do that here, but might be worth discussing if it should be changed in other parts of the code and what it would impact.

File: packages/api/src/db/setting.ts (hunk 1 above)
Comment (penk85) [reply packages/api/src/db/setting.ts:9]: For `create` not ,because I thought its best if the restrictions cannot be updated. Perhaps there's a more concise way to set this up though.

File: packages/api/src/db/setting.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+export interface UpdateSettingArgs {
+  id: string
+  input: SettingInput<any>
+}
+
+export interface DeleteSettingArgs {
+  id: string
+}
+
+export interface SettingRestriction {
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/api/src/db/setting.ts:29]: I had backend in mind here because I felt frontend validation shouldn't be enough for changing setting values, is this correct? I have a check in the mutation file.

File: packages/editor/src/client/api/setting.graphql [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            allowedValues
+        }
+    }
+}
+
+mutation UpdateSetting($id: String!, $input: SettingInput!) {
+  updateSetting(id: $id, input: $input) {
+    value
+  }
+}
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/editor/src/client/api/setting.graphql:44]: Its for the frontend access to the graphql so I haven't had use for it. In fact I'm not sure if its necessary at all because settings would need a migration to add or remove. Maybe we can discuss if this is the ideal approach.

**Comment by tomaszdurka** [comment]
@penk85 is this one close to be ready? I think it would be great to have some pairs of eyes on it.

@michael-scheurer @elias-summermatter can you check this one - it's mostly feature for operators so would be great to have your feedback here sooner than later. If you already talked it over ignore this one please. Thanks.

**Comment by penk85** [comment]
@tomaszdurka from the structure, yes, it's basically all in place, and feedback would be great. I'm still going to need to implement each setting though which I expect can take a bit time.

File: packages/api-db-mongodb/src/db/schema.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api-db-mongodb/src/db/schema.ts:443]: If we set T = unknown we don't have to set it as any below

[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api-db-mongodb/src/db/schema.ts:443]: We could also use the SettingName and map it to a respective type, if you want to do that and need help, let me know!

File: packages/api/src/db/setting.ts (hunk 1 above)
Comment (Itrulia) [reply packages/api/src/db/setting.ts:15]: We could do `export type SeetingInput<T = unknown> = Pick<Setting<T>,'value'>`.

File: packages/api/src/graphql/mutation.private.ts [hunk 7]
```diff
@@ -1152,8 +1155,43 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+
+    updateSetting: {
+      type: GraphQLSetting,
+      args: {
+        id: {type: GraphQLNonNull(GraphQLString)},
+        input: {type: GraphQLNonNull(GraphQLSettingInput)}
+      },
+
+      async resolve(root, {id, input}, {authenticate, dbAdapter, loaders}) {
+        const fullSetting = await loaders.settingsByID.load(id)
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.private.ts:1171]: Alternatively could throw a `NotFound` error too

File: packages/api/src/graphql/mutation.private.ts [hunk 8]
```diff
@@ -1152,8 +1155,43 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+        const restrictions = fullSetting?.settingRestriction
+
+        const {roles} = authenticate()
+        authorise(CanUpdateSetting, roles)
+
+        if (
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.private.ts:1177]: Can we break this `if` statement down into smaller steps/have variables for it? It's hard to follow inline like this.

I agree with @antkiewiczk we should start using a library like `joi` or `yup`

File: packages/api/src/graphql/mutation.public.ts [hunk 9]
```diff
@@ -148,20 +149,22 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
...
-      async resolve(_, {input}, {optionalAuthenticateUser, dbAdapter, challenge}) {
+      async resolve(_, {input}, {optionalAuthenticateUser, dbAdapter, challenge, loaders}) {
         const user = optionalAuthenticateUser()
         let authorType = CommentAuthorType.VerifiedUser
         const commentLength = countRichtextChars(0, input.text)
 
         if (commentLength > MAX_COMMENT_LENGTH) {
           throw new CommentLengthError()
         }
+        const guestCanCommentSetting = await dbAdapter.setting.getSetting(
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.public.ts:160]: I would redo this a tiny bit for readability:

ala:
```typescript
        const canGuestsComment = (await dbAdapter.setting.getSetting(
          SettingName.ALLOW_GUEST_COMMENTING
        ))?.value
```

The if below would then read:

```typescript
if (!canGuestsComment) throw new AnonymousCommentsDisabledError()
```

File: packages/editor/src/client/api/index.ts [hunk 10]
```diff
@@ -20,6 +20,7 @@ export type Scalars = {
   Color: string;
   RichText: Node[];
   Slug: string;
+  value: any;
```
Comment (Itrulia) [inline packages/editor/src/client/api/index.ts:23]: Something doesn't seem right here I think, this probably should be uppercase, no?
Something went wrong with generation?

File: packages/editor/src/client/api/index.ts [hunk 11]
```diff
@@ -1049,6 +1051,12 @@ export type MutationRequestChangesOnCommentArgs = {
   rejectionReason: CommentRejectionReason;
 };
 
+
+export type MutationUpdateSettingArgs = {
+  id: Scalars['String'];
```
Comment (Itrulia) [inline packages/editor/src/client/api/index.ts:1056]: This should be generated as ID not string I think?

File: packages/editor/src/client/api/setting.graphql (hunk 6 above)
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/editor/src/client/api/setting.graphql:44]: Yeah imo settings should only be deleted/added via migrations

File: packages/editor/src/client/locales/en.json [hunk 12]
```diff
@@ -560,7 +560,17 @@
       "subscriptions": "Subscriptions",
       "userRoles": "User Roles",
       "users": "Users",
-      "usersAndMembers": "Users & Members"
+      "usersAndMembers": "Users & Members",
+      "settings": "Settings",
+      "settingsPanel": {
```
Comment (Itrulia) [inline packages/editor/src/client/locales/en.json:565]: Probably should also add them to de/fr

File: packages/editor/src/client/panel/settingsPanel.tsx [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+import {Button, Drawer, Form, InputNumber, Notification, Panel, toaster, Toggle} from 'rsuite'
+import {Setting, SettingName, useSettingListQuery, useUpdateSettingMutation} from '../api'
+
+export type SettingInputProps = {
+  id?: string
+  value: any
+}
+
+export function SettingsPanel() {
+  /*
```
Comment (Itrulia) [inline packages/editor/src/client/panel/settingsPanel.tsx:13]: Can we delete this? :)

File: packages/api/src/graphql/mutation.public.ts (hunk 9 above)
Comment (penk85) [reply packages/api/src/graphql/mutation.public.ts:160]: I can't access value in the first statement because it's still a promise.

File: packages/editor/src/client/api/index.ts (hunk 10 above)
Comment (penk85) [reply packages/editor/src/client/api/index.ts:23]: Good catch

File: packages/api/src/graphql/mutation.public.ts (hunk 9 above)
Comment (Itrulia) [reply packages/api/src/graphql/mutation.public.ts]: By wrapping the await in braces you can

**Review by camilleory — DISMISSED** [review]
🥇 Nice job !

File: packages/api-db-mongodb/src/db/setting.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    const settings = await this.settings.find({_id: {$in: ids}}).toArray()
+    const settingMap = Object.fromEntries(
+      settings.map(({_id: id, ...setting}) => [id, {id, ...setting}])
+    )
+    return ids.map(id => settingMap[id] ?? null)
+  }
+
+  async updateSettingList(args: UpdateSettingArgs[]): Promise<OptionalSetting[]> {
+    const settings: OptionalSetting[] = []
+    for (const v of args) {
```
Comment (Itrulia) [inline packages/api-db-mongodb/src/db/setting.ts:57]: could also use map:

```typescript
return Promise.all(args.map(async ({id, value: val}) => {
  const {value} = await ....
  
  return {_id: outID, ...setting}
})
```

File: packages/api/src/context.ts [hunk 15]
```diff
@@ -296,8 +305,15 @@ export async function contextFromRequest(
               console.error(peer)
               return null
             }
-
-            const fetcher = createFetcher(url.resolve(peer.hostURL, 'admin'), peer.token)
+            // TODO fix types
```
Comment (Itrulia) [inline packages/api/src/context.ts:308]: Can we get that fixed before merging?

File: packages/api/src/context.ts [hunk 16]
```diff
@@ -296,8 +305,15 @@ export async function contextFromRequest(
...
+            const peerTimeout = Number(
```
Comment (Itrulia) [inline packages/api/src/context.ts:309]: don't we want to do a parseInt/+ here?

**Review by Itrulia — CHANGES_REQUESTED** [review]

File: packages/api-db-mongodb/src/db/setting.ts (hunk 14 above)
Comment (penk85) [reply packages/api-db-mongodb/src/db/setting.ts]: `Promise.all` :ok_hand:

File: packages/api/src/context.ts (hunk 16 above)
Comment (penk85) [reply packages/api/src/context.ts:308]: Regarding type fix. I'm not sure if there's a way to handle how the expected type behaves here. It would always be a number but I get the warning because its expecting unknown. Is there a cleaner way to handle this?

File: packages/api/src/context.ts [hunk 17]
```diff
@@ -269,8 +276,10 @@ export async function contextFromRequest(
               console.error(peer)
               return null
             }
-
-            const fetcher = createFetcher(peer.hostURL, peer.token)
+            const peerTimeout = Number(
+              (await dbAdapter.setting.getSetting(SettingName.PEERING_TIMEOUT_MS))?.value
```
Comment (michael-scheurer) [inline packages/api/src/context.ts:280]: what happens, if the value is undefined or null? what would Numer() return? And how would the createFetcher react?

File: packages/api/src/db/setting.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
+export type Setting<T = unknown> = {
+  id: string
+
```
Comment (michael-scheurer) [inline packages/api/src/db/setting.ts:3]: unnecessary white line

File: packages/api/src/db/setting.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+  name: SettingName
+  value: T | null
+  settingRestriction?: SettingRestriction
+}
+
+export enum SettingName {
+  DEFAULT = '',
```
Comment (michael-scheurer) [inline packages/api/src/db/setting.ts:10]: What is that? Empty string seems dangerous to me. When do you use it?

File: packages/api/src/graphql/mutation.private.ts [hunk 20]
```diff
@@ -323,9 +327,13 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
         const user = await dbAdapter.user.getUser(email)
         if (!user) throw new NotFound('User', email)
+        const jwtExpires = (
+          await dbAdapter.setting.getSetting(SettingName.SEND_LOGIN_JWT_EXPIRES_MIN)
+        )?.value
         const token = generateJWT({
           id: user.id,
-          expiresInMinutes: parseInt(process.env.SEND_LOGIN_JWT_EXPIRES_MIN as string)
+          expiresInMinutes:
+            Number(jwtExpires) ?? parseInt(process.env.SEND_LOGIN_JWT_EXPIRES_MIN as string)
```
Comment (michael-scheurer) [inline packages/api/src/graphql/mutation.private.ts:336]: why do you make it backwards compatible here but not the others?

File: packages/api/src/graphql/mutation.private.ts [hunk 21]
```diff
@@ -352,11 +360,15 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+        const jwtExpires = (
+          await dbAdapter.setting.getSetting(SettingName.SEND_LOGIN_JWT_EXPIRES_MIN)
+        )?.value
         const user = await dbAdapter.user.getUser(email)
         if (!user) throw new NotFound('User', email)
         const token = generateJWT({
           id: user.id,
-          expiresInMinutes: parseInt(process.env.SEND_LOGIN_JWT_EXPIRES_MIN as string)
+          expiresInMinutes:
+            Number(jwtExpires) ?? parseInt(process.env.SEND_LOGIN_JWT_EXPIRES_MIN as string)
```
Comment (michael-scheurer) [inline packages/api/src/graphql/mutation.private.ts:371]: idem

File: packages/editor/src/client/locales/de.json [hunk 22]
```diff
@@ -889,6 +890,28 @@
         "date": "Deaktivierung"
       }
     },
+    "errorMessages": {
+      "noFirstNameErrorMessage": "Please enter a firstname",
+      "noNameErrorMessage": "Please enter a name",
```
Comment (michael-scheurer) [inline packages/editor/src/client/locales/de.json:895]: Bitte Namen eingeben

File: packages/editor/src/client/locales/de.json [hunk 23]
```diff
@@ -889,6 +890,28 @@
...
+      "noEmailErrorMessage": "Please enter an email address",
+      "invalidEmailErrorMessage": "Please enter a valid email address",
```
Comment (michael-scheurer) [inline packages/editor/src/client/locales/de.json:897]: Bitte eine gültige E-Mail eingeben

File: packages/editor/src/client/locales/de.json [hunk 24]
```diff
@@ -889,6 +890,28 @@
...
+      "noUrlErrorMessage": "Please enter a url",
+      "invalidUrlErrorMessage": "Please enter a valid url",
+      "noImageErrorMessage": "Please enter an image",
+      "noCallToActionTextErrorMessage": "Please enter a call to action text",
+      "noCallToActionImageErrorMessage": "Please enter a call to action image",
+      "noAmountErrorMessage": "Default amount will be set to 5.00 CHF if empty",
```
Comment (michael-scheurer) [inline packages/editor/src/client/locales/de.json:903]: Der Betrag ist 5 CHF, wenn Du nichts eingibst.

File: packages/editor/src/client/locales/de.json [hunk 25]
```diff
@@ -889,6 +890,28 @@
...
+      "noPasswordErrorMessage": "Please enter a password",
+      "passwordTooShortErrorMessage": "Password must be at least 8 characters long",
+      "noPaymentProviderErrorMessage": "Please enter a payment provider",
+      "noMemberPlanErrorMessage": "Please enter a member plan",
+      "noPaymentPeriodicityErrorMessage": "Please enter a payment periodicity",
+      "noPaymentMethodErrorMessage": "Please enter a paymentMethod",
+      "noUserErrorMessage": "Please select a user",
+      "noTokenErrorMessage": "Please enter a token",
+      "invalidRange": "ungültiger Wert. Bitte Zahl zwischen {{min}} und {{max}} eingeben.",
```
Comment (michael-scheurer) [inline packages/editor/src/client/locales/de.json:912]: Ungültiger Wert. Bitte Zahl zwischen {{min}} und {{max}} eingeben.

File: packages/api/src/context.ts (hunk 16 above)
Comment (Itrulia) [reply packages/api/src/context.ts:303]: If it's already coming as number you should type it as one (value as number). parseInt, + and Number all have runtime changes.

using Number() has the side effect that strings that look like other bases, will not be in base 10. So parseInt/+ is preferred.

File: packages/api/src/context.ts (hunk 17 above)
Comment (penk85) [reply packages/api/src/context.ts:275]: I can use a fallback to the env variable, but that's not guaranteed to be defined either. Would it be better to throw an error or to set a default value?

File: packages/api/src/context.ts (hunk 16 above)
Comment (penk85) [reply packages/api/src/context.ts:303]: And helpful to know about parseInt vs Number() :+1:

**Review by Itrulia — APPROVED** [review]

File: packages/api/src/graphql/mutation.private.ts [hunk 26]
```diff
@@ -1144,8 +1157,51 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+            throw new NotFound('setting', name)
+          }
+
+          const restrictions = fullSetting?.settingRestriction
+
+          const allowedValues = restrictions?.allowedValues
+            ? restrictions?.allowedValues?.includes(String(val))
+            : true
+          const inputLength =
+            restrictions?.inputLength !== undefined
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.private.ts:1188]: This wouldn't catch null/''/0

```suggestion
            restrictions?.inputLength
```

File: packages/api/src/graphql/mutation.private.ts [hunk 27]
```diff
@@ -1144,8 +1157,51 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+              ? restrictions.inputLength >= String(val).length
+              : true
+          const maxValue =
+            restrictions?.maxValue !== undefined ? restrictions.maxValue >= Number(val) : true
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.private.ts:1192]:
```suggestion
            restrictions?.maxValue ? restrictions.maxValue >= Number(val) : true
```

File: packages/api/src/graphql/mutation.private.ts [hunk 28]
```diff
@@ -1144,8 +1157,51 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+          const minValue =
+            restrictions?.minValue !== undefined ? restrictions.minValue <= Number(val) : true
```
Comment (Itrulia) [inline packages/api/src/graphql/mutation.private.ts:1194]:
```suggestion
            restrictions?.minValue ? restrictions.minValue <= Number(val) : true
```

File: packages/api/src/graphql/mutation.private.ts [hunk 29]
```diff
@@ -1144,8 +1156,51 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+        value.map(async ({id, value: val}: UpdateSettingArgs) => {
+          const fullSetting = await loaders.settingsByID.load(id)
+
+          if (!fullSetting) {
+            throw new NotFound('setting', id)
+          }
+
+          const restrictions = fullSetting?.settingRestriction
+
+          const allowedValues = restrictions?.allowedValues
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api/src/graphql/mutation.private.ts:1184]: I would move all validations in the API. The issue with validations in the database layer is that you handle possible incorrect data before you pass it to the database.

Types of GraphQL can have validations added to them (isTypeOf for Objects or Scalar for Primitives). This is advocated for by the Apollo Documentation.

**Review by camilleory — APPROVED** [review]

File: packages/editor/src/client/locales/de.json [hunk 30]
```diff
@@ -889,6 +890,28 @@
         "date": "Deaktivierung"
       }
     },
+    "errorMessages": {
+      "noFirstNameErrorMessage": "Please enter a firstname",
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/editor/src/client/locales/de.json:894]: These were just placeholders because the validation branch only had English and I wanted the keys to match.

File: packages/api/src/graphql/mutation.private.ts [hunk 31]
```diff
@@ -1144,8 +1156,51 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
+
+    updateSettingList: {
+      type: GraphQLList(GraphQLSetting),
+      args: {
+        value: {type: GraphQLList(GraphQLUpdateSettingArgs)}
+      },
+
+      async resolve(root, {value}, {authenticate, dbAdapter, loaders}) {
+        const {roles} = authenticate()
+        authorise(CanUpdateSetting, roles)
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/api/src/graphql/mutation.private.ts]: Yes, tested. Will return an error if not authorized.

**Review by Itrulia — APPROVED** [review]

**Review by michael-scheurer — DISMISSED** [review]
I pushed a fix which prevents a potential runtime error. imo u can merge.

**Review by Itrulia — APPROVED** [review]

**Review by michael-scheurer — APPROVED** [review]
